File: examples/TimeIntegers.py
#!/usr/bin/env python3
# TimeIntegers.py: time integer draws and permutations (ns/value) in Python.

import argparse
import time
from dataclasses import dataclass
from typing import Callable, List

import numpy as np


@dataclass
class IntRangeSpec:
  m: int
  n: int
  label: str


@dataclass
class U8Spec:
  bound: int
  label: str


@dataclass
class PermSpec:
  n: int
  label: str


@dataclass
class SampleSpec:
  n: int
  k: int
  label: str


def compute_reps(chunk: int) -> int:
  return max(1, 1_000_000 // chunk)


def time_int_range(rng: np.random.Generator,
                   chunk: int,
                   bench_time: float,
                   m: int,
                   n: int) -> float:
  reps = compute_reps(chunk)
  calls = 0
  t0 = time.perf_counter()
  t = t0
  while t - t0 < bench_time:
    for _ in range(reps):
      rng.integers(m, n + 1, size=chunk, dtype=np.int32)
    calls += reps
    t = time.perf_counter()
  total_vals = calls * chunk
  if total_vals <= 0:
    return float("nan")
  return 1e9 * (t - t0) / total_vals


def time_uint8_bound(rng: np.random.Generator,
                     chunk: int,
                     bench_time: float,
                     bound: int) -> float:
  reps = compute_reps(chunk)
  calls = 0
  t0 = time.perf_counter()
  t = t0
  while t - t0 < bench_time:
    for _ in range(reps):
      rng.integers(0, bound, size=chunk, dtype=np.uint8)
    calls += reps
    t = time.perf_counter()
  total_vals = calls * chunk
  if total_vals <= 0:
    return float("nan")
  return 1e9 * (t - t0) / total_vals


def time_uint64_bound(rng: np.random.Generator,
                      chunk: int,
                      bench_time: float,
                      bound: int) -> float:
  reps = compute_reps(chunk)
  calls = 0
  t0 = time.perf_counter()
  t = t0
  while t - t0 < bench_time:
    for _ in range(reps):
      rng.integers(0, bound, size=chunk, dtype=np.uint64)
    calls += reps
    t = time.perf_counter()
  total_vals = calls * chunk
  if total_vals <= 0:
    return float("nan")
  return 1e9 * (t - t0) / total_vals


def time_perm(rng: np.random.Generator,
              bench_time: float,
              n: int) -> float:
  reps = max(1, 100_000 // n)
  calls = 0
  t0 = time.perf_counter()
  t = t0
  while t - t0 < bench_time:
    for _ in range(reps):
      rng.permutation(n)
    calls += reps
    t = time.perf_counter()
  total_vals = calls * n
  if total_vals <= 0:
    return float("nan")
  return 1e9 * (t - t0) / total_vals


def time_sample(rng: np.random.Generator,
                bench_time: float,
                n: int,
                k: int) -> float:
  reps = max(1, 100_000 // n)
  calls = 0
  t0 = time.perf_counter()
  t = t0
  while t - t0 < bench_time:
    for _ in range(reps):
      rng.choice(n, size=k, replace=False)
    calls += reps
    t = time.perf_counter()
  total_vals = calls * k
  if total_vals <= 0:
    return float("nan")
  return 1e9 * (t - t0) / total_vals


def make_int_ranges() -> List[IntRangeSpec]:
  return [
    IntRangeSpec(0, 2, "[0,2]"),
    IntRangeSpec(1, 10, "[1,10]"),
    IntRangeSpec(1, 100, "[1,100]"),
    IntRangeSpec(1, 1000, "[1,1000]"),
    IntRangeSpec(1, 10000, "[1,10000]"),
    IntRangeSpec(1, 100000, "[1,100000]"),
    IntRangeSpec(1, 1000000, "[1,1000000]"),
  ]


def main() -> None:
  parser = argparse.ArgumentParser(
    description="Time integer draws and permutations (ns/value)")
  parser.add_argument("-t", type=float, default=0.1, dest="bench_time",
                      help="benchmark time per case (seconds)")
  parser.add_argument("-c", type=int, default=4096, dest="chunk",
                      help="chunk size (values per call)")
  parser.add_argument("-s", type=int, default=7, dest="seed",
                      help="random seed")
  args = parser.parse_args()

  rng = np.random.default_rng(args.seed)

  int_ranges = make_int_ranges()
  u8_specs = [U8Spec(2, "bound 2"), U8Spec(10, "bound 10")]
  perm_specs = [PermSpec(100, "100"), PermSpec(100000, "100000")]
  sample_specs = [
    SampleSpec(1000, 10, "1000/10"),
    SampleSpec(1000, 499, "1000/499"),
    SampleSpec(1000, 501, "1000/501"),
    SampleSpec(1000, 990, "1000/990"),
  ]

  # Warm up.
  for spec in int_ranges:
    rng.integers(spec.m, spec.n + 1, size=16, dtype=np.int32)
  for spec in u8_specs:
    rng.integers(0, spec.bound, size=16, dtype=np.uint8)
  rng.integers(0, (1 << 64) // 3, size=16, dtype=np.uint64)
  for spec in perm_specs:
    rng.permutation(spec.n)
  for spec in sample_specs:
    rng.choice(spec.n, size=spec.k, replace=False)

  print("time per value:   ns/value")
  print(f"bench_time:       {args.bench_time:.3f} s per case")
  print(f"chunk:            {args.chunk}")
  print("\n%-14s %8s" % ("int range", "ns/value"))
  for spec in int_ranges:
    ns = time_int_range(rng, args.chunk, args.bench_time, spec.m, spec.n)
    print("%-14s %8.2f" % (spec.label, ns))
  print("\n%-14s %8s" % ("uint8", "ns/value"))
  for spec in u8_specs:
    ns = time_uint8_bound(rng, args.chunk, args.bench_time, spec.bound)
    print("%-14s %8.2f" % (spec.label, ns))
  print("\n%-14s %8s" % ("uint64", "ns/value"))
  ns_u64 = time_uint64_bound(rng, args.chunk, args.bench_time, (1 << 64) // 3)
  print("%-14s %8.2f" % ("UINT64_MAX/3", ns_u64))
  print("\n%-14s %10s" % ("perm n", "ns/value"))
  for spec in perm_specs:
    ns = time_perm(rng, args.bench_time, spec.n)
    print("%-14s %10.2f" % (spec.label, ns))
  print("\n%-14s %10s" % ("sample", "ns/value"))
  for spec in sample_specs:
    ns = time_sample(rng, args.bench_time, spec.n, spec.k)
    print("%-14s %10.2f" % (spec.label, ns))


if __name__ == "__main__":
  main()

File: Randompack.jl/examples/TimeIntegers.jl
# TimeIntegers.jl
# Time integer draws and permutations (ns/value) in Julia

using Random
using Printf

struct IntRangeSpec
  m::Int
  n::Int
  label::String
end

struct LongLongSpec
  m::Int64
  n::Int64
  label::String
end

struct U8Spec
  bound::Int
  label::String
end

struct PermSpec
  n::Int
  label::String
end

function compute_reps(chunk::Int)
  return max(1, 1_000_000 ÷ chunk)
end

function time_int_range!(rng, chunk::Int, bench_time::Float64, m::Int, n::Int)
  reps = compute_reps(chunk)
  buf = Vector{Int32}(undef, chunk)
  calls = 0
  t0 = time_ns()
  t = t0
  limit = Int(round(bench_time * 1e9))
  while (t - t0) < limit
    for _ = 1:reps
      rand!(rng, buf, m:n)
    end
    calls += reps
    t = time_ns()
  end
  total = calls * chunk
  total == 0 && return NaN
  return (t - t0) / total
end

function time_long_long_range!(rng, chunk::Int, bench_time::Float64, m::Int64, n::Int64)
  reps = compute_reps(chunk)
  buf = Vector{Int64}(undef, chunk)
  calls = 0
  t0 = time_ns()
  t = t0
  limit = Int(round(bench_time * 1e9))
  while (t - t0) < limit
    for _ = 1:reps
      rand!(rng, buf, m:n)
    end
    calls += reps
    t = time_ns()
  end
  total = calls * chunk
  total == 0 && return NaN
  return (t - t0) / total
end

function time_uint8_bound!(rng, chunk::Int, bench_time::Float64, bound::Int)
  reps = compute_reps(chunk)
  buf = Vector{UInt8}(undef, chunk)
  calls = 0
  t0 = time_ns()
  t = t0
  limit = Int(round(bench_time * 1e9))
  while (t - t0) < limit
    for _ = 1:reps
      rand!(rng, buf, UInt8(0):UInt8(bound-1))
    end
    calls += reps
    t = time_ns()
  end
  total = calls * chunk
  total == 0 && return NaN
  return (t - t0) / total
end

function time_uint64_bound!(rng, chunk::Int, bench_time::Float64, bound::UInt64)
  reps = compute_reps(chunk)
  buf = Vector{UInt64}(undef, chunk)
  calls = 0
  t0 = time_ns()
  t = t0
  limit = Int(round(bench_time * 1e9))
  while (t - t0) < limit
    for _ = 1:reps
      rand!(rng, buf, UInt64(0):bound-1)
    end
    calls += reps
    t = time_ns()
  end
  total = calls * chunk
  total == 0 && return NaN
  return (t - t0) / total
end

function time_perm!(rng, bench_time::Float64, n::Int)
  reps = max(1, 100_000 ÷ n)
  buf = collect(1:n)
  calls = 0
  t0 = time_ns()
  t = t0
  limit = Int(round(bench_time * 1e9))
  while (t - t0) < limit
    for _ = 1:reps
      shuffle!(rng, buf)
    end
    calls += reps
    t = time_ns()
  end
  total = calls * n
  total == 0 && return NaN
  return (t - t0) / total
end


function make_int_ranges()
  return [
    IntRangeSpec(1, 3, "1-3"),
    IntRangeSpec(1, 20, "1-20"),
    IntRangeSpec(1, 1000, "1-1000"),
    IntRangeSpec(1, 100000, "1-1e5"),
    IntRangeSpec(1, 10000000, "1-1e7"),
    IntRangeSpec(1, 1000000000, "1-1e9"),
  ]
end

function warmup!(seconds::Float64)
  rng = Xoshiro()
  t0 = time_ns()
  limit = Int(round(seconds * 1e9))
  buf32 = Vector{Int32}(undef, 1024)
  buf64 = Vector{Int64}(undef, 1024)
  buf8 = Vector{UInt8}(undef, 1024)
  perm_buf = collect(1:1000)
  while (time_ns() - t0) < limit
    rand!(rng, buf32, 1:1000)
    rand!(rng, buf64, Int64(1):Int64(1000000))
    rand!(rng, buf8, UInt8(0):UInt8(9))
    shuffle!(rng, perm_buf)
  end
  return nothing
end

function main()
  bench_time = 0.2
  chunk = 4096
  seed = nothing
  i = 1
  while i <= length(ARGS)
    arg = ARGS[i]
    if arg == "-h" || arg == "--help"
      println("Usage: julia TimeIntegers.jl [-t sec] [-c chunk] [-s seed]")
      println("  -t sec     time per case in seconds (default: 0.2)")
      println("  -c chunk   number of draws per call (default: 4096)")
      println("  -s seed    fixed seed (default random seed per case)")
      return
    elseif arg == "-t"
      if i == length(ARGS)
        error("missing value for -t")
      end
      i += 1
      bench_time = parse(Float64, ARGS[i])
      if bench_time <= 0
        error("time per case must be positive")
      end
    elseif arg == "-c"
      if i == length(ARGS)
        error("missing value for -c")
      end
      i += 1
      chunk = parse(Int, ARGS[i])
      if chunk <= 0
        error("chunk must be positive")
      end
    elseif arg == "-s"
      if i == length(ARGS)
        error("missing value for -s")
      end
      i += 1
      seed = parse(Int, ARGS[i])
    else
      error("Unknown option: $arg")
    end
    i += 1
  end

  int_ranges = make_int_ranges()
  u8_specs = [U8Spec(2, "bound 2"), U8Spec(10, "bound 10")]
  perm_specs = [PermSpec(100, "100"), PermSpec(100000, "100000")]
  ll_ranges = [
    LongLongSpec(Int64(1), Int64(10), "1-10"),
    LongLongSpec(Int64(1), Int64(1000), "1-1e3"),
    LongLongSpec(Int64(1), Int64(1000000), "1-1e6"),
    LongLongSpec(Int64(1), Int64(10000000000), "1-1e10"),
    LongLongSpec(Int64(1), Int64(1000000000000000000), "1-1e18"),
  ]

  warmup!(0.1)

  println("time per value:   ns/value")
  @printf("bench_time:       %.3f s per case\n", bench_time)
  @printf("chunk:            %d\n", chunk)
  @printf("\n%-14s %8s\n", "int range", "ns/value")
  for spec in int_ranges
    case_seed = seed === nothing ? rand(RandomDevice(), 0:2^31-1) : seed
    rng = Xoshiro(case_seed)
    ns = time_int_range!(rng, chunk, bench_time, spec.m, spec.n)
    @printf("%-14s %8.2f\n", spec.label, ns)
  end
  @printf("\n%-14s %8s\n", "long long", "ns/value")
  for spec in ll_ranges
    case_seed = seed === nothing ? rand(RandomDevice(), 0:2^31-1) : seed
    rng = Xoshiro(case_seed)
    ns = time_long_long_range!(rng, chunk, bench_time, spec.m, spec.n)
    @printf("%-14s %8.2f\n", spec.label, ns)
  end
  @printf("\n%-14s %8s\n", "uint8", "ns/value")
  for spec in u8_specs
    case_seed = seed === nothing ? rand(RandomDevice(), 0:2^31-1) : seed
    rng = Xoshiro(case_seed)
    ns = time_uint8_bound!(rng, chunk, bench_time, spec.bound)
    @printf("%-14s %8.2f\n", spec.label, ns)
  end
  @printf("\n%-14s %8s\n", "uint64", "ns/value")
  case_seed = seed === nothing ? rand(RandomDevice(), 0:2^31-1) : seed
  rng = Xoshiro(case_seed)
  ns_u64 = time_uint64_bound!(rng, chunk, bench_time, typemax(UInt64) ÷ 3)
  @printf("%-14s %8.2f\n", "UINT64_MAX/3", ns_u64)
  @printf("\n%-14s %10s\n", "perm n", "ns/value")
  for spec in perm_specs
    case_seed = seed === nothing ? rand(RandomDevice(), 0:2^31-1) : seed
    rng = Xoshiro(case_seed)
    ns = time_perm!(rng, bench_time, spec.n)
    @printf("%-14s %10.2f\n", spec.label, ns)
  end
end

main()
